### ISSUE TITLE
The Reader and Table fields are missed in the System.Resources.ResourceSet class

### ISSUE THREAD

**Issue by 87Alex** [dotnet/corefx#36882]
* .NET Core Version: (e.g. 3.0 Preview1, or daily build number, use `dotnet --info`)
* Have you experienced this same bug with .NET Framework?: Yes/No
->No

**Problem description:**
The Reader and Table fields are not available in the System.Resources.ResourceSet class descendant.

**Actual behavior:** 
The Reader and Table fields are missed in the System.Resources.ResourceSet class.

**Expected behavior:**
The Reader and Table protected fields are available in the System.Resources.ResourceSet class as with .NET Framework.

**Minimal repro:**
[ConsoleApp1.zip](https://github.com/dotnet/corefx/files/3080668/ConsoleApp1.zip)

**Comment by tarekgh** [comment]
@87Alex in your example, why you need to subclass the ResourceSet at all? would it be simpler to create the ResourceSet object directly using the IResourceReader (which already implemented by XmlResourceReader)? 

```C#
var resourceSet = new ResourceSet(new XmlResourceReader());
```

I am not seeing any benefit of sub-classing the ResourceSet not exposing these fields.

CC @jkotas @danmosemsft @krwq

**Comment by 87Alex** [comment]
Hello,
It is just a simple demonstrating project, the real descendant is more complex:
1) In our descendant, we override the ReadResources method with a custom logic and we need access to the Reader and Table fields.
2) We have a constructor with additional settings which are used in the ReadResources method. When we create the ResourceSet object directly using the IResourceReader, it immediately calls the ReadResources() method. Our additional settings are not yet initialized at this moment and our logic works incorrectly.

**Comment by tarekgh** [comment]
@87Alex I am trying to figure out if the Reader and Table fields are really needed in your scenario or your code can have some fix to get rid of such dependency. 

> When we create the ResourceSet object directly using the IResourceReader, it immediately calls the ReadResources() method. 

I am wondering, can't you just delay creating the ResourecSet till you do finish the needed initialization. 

We are trying to not expose these fields because it is not right design. We are trying to figure out if there is any scenario cannot be achieved without exposing these fields.

**Comment by 87Alex** [comment]
Hello,
I understand the reason why you deleted these fields and agree that it is not the right design. I researched the ResourceSet class sources in search of ways to change our code but did not find a solution. As I mentioned before, we override the ReadResources method, and fill the Table with custom values. If the Table field is not available, then the virtual parameter of the ReadResources method makes no sense because there is no other way to fill the Table.
As for Reader field, I think we can store it to local variable in our constructor and use it in the future and we can postpone ReadResources method until our constructor with required settings is created.
To summarize, in our descendant, we need some way to get and update the Table field, because we need to check its contents while adding new items, rather than piping items sequentially.
I attached the overridden ReadResources method of our descendant.
[ReadResources.txt](https://github.com/dotnet/corefx/files/3090229/ReadResources.txt)

**Comment by tarekgh** [comment]
Thanks @87Alex for the details.

Is it possible you don't subclass ResourceSet at all. Then just use  the constructor:

```C#
    var resourceSet = new ResourceSet(new CustomResourceReader(...));
```

When this constructor get called, this constructor will call ```CustomResourceReader.GetEnumerator()``` to fill the internal and not exposed Table field. So in your implementation of ```CustomResourceReader.GetEnumerator()```, you can initialize some private table the way you have done [ReadResources.txt](https://github.com/dotnet/corefx/files/3090229/ReadResources.txt) and after finishing this initialization, you can return the enumerator on this private table. Doing that, will ensure your initialization logic still work as it used before. Sorry if I am missing anything here. 

The question now, after finishing the ResourceSet creation (and filling the table with the desired resources), does your scenario need to add more entries to the resource table? and if so, how frequently this will happen?

**Comment by 87Alex** [comment]
Hello,
We don't need to add more entries to the resource table but we need to clear the table and read resources again when the application model used in the ResourceSet is changed. If we move our logic from the ResourceSet descendant to a custom reader, it may break the current logic and requires additional time to rewrite and test it.

**Comment by tarekgh** [comment]
> read resources again when the application model used in the ResourceSet is changed

You can recreate the ResourceSet object at that time.

> it may break the current logic 

I don't know know exactly your situation but I am not seeing anything blocking to have this nicely work.

> requires additional time to rewrite and test it.

I think it is worth to invest on that.

Per that, as I am mentioned before we are not inclining to expose these fields and there is a workaround for your scenario which we recommend to do. I am closing this issue, but feel free to respond back with any question. Thanks again for your report and sharing your scenario.
